src/inference/ais.jl

```diff
@@ -0,0 +1,339 @@
+    "log of the average of the particle weights: estimator of the log evidence - computed in sample_end!"
+    final_logevidence  ::  F
+end
+
+AISState(model::Model) = AISState(VarInfo(model), [], 0.0)
+
+# A.3. Sampler constructor: same as for vanilla IS
+
+"""
+    Sampler(alg::AIS, model::Model, s::Selector)
+
+Sampler constructor: similar to vanilla IS.
+"""
+function Sampler(alg::AIS, model::Model, s::Selector)
+    @assert length(alg.schedule) == length(alg.proposal_kernels)
+    info = Dict{Symbol, Any}()
+    state = AISState(model)
+    return Sampler(alg, info, s, state)
+end
+
+## B. Implement AbstractMCMC
+
+# B.1. new transition type AISTransition, with an additional attribute accum_logweight
+
+"""
+    AISTransition{T, F<:AbstractFloat}
+
+AIS-specific Transition struct. 
+
+Necessary because we care both about a particle's weight (accum_logweight) and the logjoint density evaluated at its final position (lp).
+"""
+struct AISTransition{T, F<:AbstractFloat}
+    "parameter"
+    θ  :: T
+    "logjoint evaluated at θ"
+    lp :: F
+    "logarithm of the particle's AIS weight - accumulated during annealing run"
+    accum_logweight :: F
+end
+
+# idk what this function is for
+function additional_parameters(::Type{<:AISTransition})
+    return [:lp, :accum_logweight]
+end
+
+
+# B.2. sample_init! function 
+
+"""
+    AbstractMCMC.sample_init!
+
+Initialize AISState attributes densitymodels and logjoint in spl.state.
+"""
+function AbstractMCMC.sample_init!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:AIS},
+    N::Integer;
+    verbose::Bool=true,
+    resume_from=nothing,
+    kwargs...
+)
+    logjoint = gen_logjoint(spl.state.vi, model, spl)
+    logprior = gen_logprior(spl.state.vi, model, spl)
+
+    spl.state.densitymodels = [] 
+
+    # densitymodels for intermediate distributions
+    for beta in spl.alg.schedule
+        log_unnorm_tempered = gen_log_unnorm_tempered(logprior, logjoint, beta)
+        densitymodel = AdvancedMH.DensityModel(log_unnorm_tempered)
+        push!(spl.state.densitymodels, densitymodel)
+    end
+
+    # densitymodel for the logjoint, ie the final target
+    final_densitymodel = AdvancedMH.DensityModel(logjoint)
+    push!(spl.state.densitymodels, final_densitymodel)
+end
+
+# B.3. step function 
+
+# TODO: memorize full path: not much to change, but check what we care about exactly
+# - all proposals?
+# - all values of accum_logweight?
+# - all values of diff_logdensity?
+# - all values of f_j(x_j)?
+
+function AbstractMCMC.step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:AIS},
+    ::Integer,
+    transition;
+    kwargs...
+)
+    # sample from prior and compute first term in accum_logweight
+    current_state, accum_logweight = prior_step(spl, model)
+
+    # for every intermediate distribution
+    for j in 1:length(spl.alg.schedule)
+        current_state, accum_logweight = intermediate_step(j, spl, current_state, accum_logweight)
```

Comment:
```suggestion
        densitymodel = spl.state.densitymodels[j]
        proposal_kernel = spl.alg.proposal_kernels[j]
        current_state, accum_logweight = intermediate_step(densitymodel, proposal_kernel, current_state, accum_logweight)
```
See my comment below on `intermediate_step`